### ISSUE TITLE
Postgres: ALTER USER/ROLE CURRENT_ROLE/SESSION_USER not parsable

### ISSUE THREAD

**Issue by publictheta** [sqlfluff/sqlfluff#4956]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

SQLFluff does not parse the following statements correctly:

```sql
-- FAIL: found unparsable section
ALTER USER CURRENT_ROLE WITH PASSWORD NULL;
ALTER USER SESSION_USER WITH PASSWORD NULL;
ALTER ROLE CURRENT_ROLE WITH PASSWORD NULL;
ALTER ROLE SESSION_USER WITH PASSWORD NULL;
```

while it parses the following statements correctly:

```sql
-- OK
ALTER USER CURRENT_USER WITH PASSWORD NULL;
ALTER USER my_name WITH PASSWORD NULL;
ALTER ROLE CURRENT_USER WITH PASSWORD NULL;
ALTER ROLE my_name WITH PASSWORD NULL;
```

Additionally, the following statement should not be parsed correctly, but it is:

```sql
-- MUST FAIL
ALTER USER ALL RENAME TO my_name;
```

### Expected Behaviour

SQLFluff should parse the above statements correctly.

The syntax of [`ALTER USER`] (an alias for [`ALTER ROLE`]) allows the following depending on the context:

`my_name` only:

```sql
ALTER USER ... RENAME TO ...;
```

`my_name` or `CURRENT_ROLE`/`CURRENT_USER`/`SESSION_USER`:

```sql
ALTER USER ... option;
ALTER USER ... WITH option;
```

`my_name`, `CURRENT_ROLE`/`CURRENT_USER`/`SESSION_USER`, or `ALL`:

```sql
ALTER USER ... SET ...;
ALTER USER ... RESET ...;
```

[`ALTER USER`]: https://www.postgresql.org/docs/current/sql-alteruser.html
[`ALTER ROLE`]: https://www.postgresql.org/docs/current/sql-alterrole.html

NOTE: The current version of PostgreSQL is 15 ([`ALTER USER`](https://www.postgresql.org/docs/15/sql-alteruser.html), [`ALTER ROLE`](https://www.postgresql.org/docs/15/sql-alterrole.html)).

### Observed Behaviour

SQLFluff reports "found unparsable section" errors as follows:

```
== [sample.sql] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'ALTER
                       | USER CURRENT_ROLE WITH PASSWORD NU...'
L:   2 | P:   1 |  PRS | Line 2, Position 1: Found unparsable section: 'ALTER
                       | USER SESSION_USER WITH PASSWORD NU...'
L:   3 | P:   1 |  PRS | Line 3, Position 1: Found unparsable section: 'ALTER
                       | ROLE CURRENT_ROLE WITH PASSWORD NU...'
L:   4 | P:   1 |  PRS | Line 4, Position 1: Found unparsable section: 'ALTER
                       | ROLE SESSION_USER WITH PASSWORD NU...'
```

From a quick look at the source code, it seems that the following parts of the grammar are relevant:

[`src/sqlfluff/dialects/dialect_postgres.py#L1560-L1568`](https://github.com/sqlfluff/sqlfluff/blob/4a9152d3acda887aed551800b8c3a8833d72bd5e/src/sqlfluff/dialects/dialect_postgres.py#L1560-L1568)

```python
    match_grammar = Sequence(
        "ALTER",
        OneOf("ROLE", "USER"),
        OneOf(Ref("RoleReferenceSegment"), "ALL"),
        OneOf(
            Sequence(
                Ref.keyword("WITH", optional=True),
                AnySetOf(
                    OneOf("SUPERUSER", "NOSUPERUSER"),
        # ...
```

It allows `my_name` or `ALL` for any context. And it also allows `CURRENT_USER` because it is not "reserved" as the comment on the other line [says].

[says]: https://github.com/sqlfluff/sqlfluff/blob/4a9152d3acda887aed551800b8c3a8833d72bd5e/src/sqlfluff/dialects/dialect_postgres.py#L2308-L2314

To fix the issue, we may need to move the role reference part to the beginning of each subsequent `Sequence`  as follows:

```python
    match_grammar = Sequence(
        "ALTER",
        OneOf("ROLE", "USER"),
        # OneOf(Ref("RoleReferenceSegment"), "ALL"),
        OneOf(
            Sequence(
                OneOf(
                    "CURRENT_ROLE",
                    "CURRENT_USER",
                    "SESSION_USER",
                    # must come last; CURRENT_USER isn't reserved:
                    Ref("RoleReferenceSegment"),
                ),
                Ref.keyword("WITH", optional=True),
            # ...
            Sequence(
                Ref("RoleReferenceSegment"),
                "RENAME",
            # ...
            Sequence(
                OneOf(
                    "CURRENT_ROLE",
                    "CURRENT_USER",
                    "SESSION_USER",
                    "ALL",
                    # must come last; CURRENT_USER isn't reserved:
                    Ref("RoleReferenceSegment"),
                ),
            # ...
```

(Sorry, I'm not familiar with the codebase, so I'm not sure if this is the proper way to fix the issue. I just wanted to
ease the burden of whoever is going to fix the issue.)

### How to reproduce

1. Create `sample.sql` with the following content:

```sql
ALTER USER CURRENT_ROLE WITH PASSWORD NULL;
ALTER USER SESSION_USER WITH PASSWORD NULL;
ALTER ROLE CURRENT_ROLE WITH PASSWORD NULL;
ALTER ROLE SESSION_USER WITH PASSWORD NULL;
```

2. Run:

```shell
sqlfluff lint sample.sql --dialect=postgres
```

### Dialect

postgres

### Version

sqlfluff, version 2.1.2
Python 3.11.4

### Configuration

(default)

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by anzelpwj** [comment]
Happy to attempt to work on this.
